lib/buffer.js

```diff
@@ -539,11 +539,29 @@ Buffer.prototype.inspect = function inspect() {
   return '<' + this.constructor.name + ' ' + str + '>';
 };
 
+function assertInteger(val, name) {
+  if (typeof val !== 'undefined' &&
+      (typeof val !== 'number' ||
+       +val !== (val >> 0))) {
+    throw new TypeError(`"${name}" argument must be an integer`);
+  }
+}
 
-Buffer.prototype.compare = function compare(b) {
+Buffer.prototype.compare = function compare(b, start, end, thisStart, thisEnd) {
   if (!(b instanceof Buffer))
     throw new TypeError('Argument must be a Buffer');
 
+  assertInteger(start, 'targetStart');
+  assertInteger(thisEnd, 'sourceEnd');
+  assertInteger(end, 'targetEnd');
+  assertInteger(thisStart, 'sourceStart');
```

Comment:
If you want to experiment moving all of this to native for performance, be my guest.